main/sdr.go

```diff
@@ -319,20 +312,96 @@ func sdrKill() {
 	}
 }
 
-// Watch for config/device changes.
-func sdrWatcher() {
-	var doSkip bool
-	rES, err := regexp.Compile("str?a?t?u?x:1090")
-	if err != nil {
-		rES = nil
-		log.Println("failed to compile ES regexp because %s", err.Error())
+func reCompile(s string) *regexp.Regexp {
+	// note , compile returns a nil pointer on error
+	r, _ := regexp.Compile(s)
+	return r
+}
+
+type regexUAT regexp.Regexp
+type regexES regexp.Regexp
+
+var rUAT = (*regexUAT)(reCompile("str?a?t?u?x:978"))
+var rES = (*regexUAT)(reCompile("str?a?t?u?x:1090"))
+
+func (r *regexUAT) hasID(serial string) bool {
+	if r == nil {
+		return strings.HasPrefix(serial, "stratux:978")
 	}
+	return (*regexp.Regexp)(r).MatchString(serial)
+}
 
-	rUAT, err := regexp.Compile("str?a?t?u?x:978")
-	if err != nil {
-		rUAT = nil
-		log.Println("failed to compile UAT regexp because %s", err.Error())
+func (r *regexES) hasID(serial string) bool {
+	if r == nil {
+		return strings.HasPrefix(serial, "stratux:1090")
+	}
+	return (*regexp.Regexp)(r).MatchString(serial)
+}
+
+func createUATDev(id int, serial string, idSet bool) error {
+	UATDev = &UAT{indexID: id, serial: serial}
+	if err := UATDev.sdrConfig(); err != nil {
+		log.Printf("UATDev.sdrConfig() failed: %s\n", err)
+		UATDev = nil
+		return err
 	}
+	UATDev.wg = &sync.WaitGroup{}
+	UATDev.idSet = idSet
+	UATDev.closeCh = make(chan int)
+	UATDev.wg.Add(1)
+	go UATDev.read()
+	return nil
+}
+
+func createESDev(id int, serial string, idSet bool) error {
+	ESDev = &ES{indexID: id, serial: serial}
+	if err := ESDev.sdrConfig(); err != nil {
+		log.Printf("ESDev.sdrConfig() failed: %s\n", err)
+		ESDev = nil
+		return err
+	}
+	ESDev.wg = &sync.WaitGroup{}
+	ESDev.idSet = idSet
+	ESDev.closeCh = make(chan int)
+	ESDev.wg.Add(1)
+	go ESDev.read()
+	return nil
+}
+
+func configDevices(count int, es_enabled, uat_enabled bool) {
+	// entry to this function is only valid when both UATDev and ESDev are nil
+	for i := 0; i < count; i++ {
+		_, _, s, err := rtl.GetDeviceUsbStrings(i)
+		if err == nil {
+			if uat_enabled && UATDev == nil && rUAT.hasID(s) {
+				createUATDev(i, s, true)
+				continue
+			}
+			if es_enabled && ESDev == nil && rES.hasID(s) {
+				createESDev(i, s, true)
+				continue
+			}
+
+			// we only get here when it's an anonymous dongle
+			if uat_enabled && UATDev == nil {
```

Comment:
This still has the bug that I fixed in #278: if i 0 is anonymous and i 1 is wired to UAT, then only 0 will be activated on UAT.

The fix is to run the loop twice: first to assign non-anonymous dongles and the second to assign anonymous dongles. Ideally a third loop would assign all dongles, in case a user erroneously assigned both dongles to :1090. This logic would also implement #239.